Improve readability of import_app

gherceg opened the pull request:
## Summary
<!--
    Provide a link to the ticket or document which prompted this change,
    Describe the rationale and design decisions.
-->
I was trying to understand this and felt overwhelmed by everything going on in this method. I still don't understand every responsibility of this method so hopefully reviewers can help suggest better names/organization to improve readability. 

Additionally if this feels like a downgrade I'm curious to hear why. 
## Feature Flag
<!-- If this is specific to a feature flag, which one? -->

## Product Description
<!-- For non-invisible changes, describe user-facing effects. -->

## Safety Assurance

- [x] Risk label is set correctly
- [x] All migrations are backwards compatible and won't block deploy
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change
- [x] If QA is part of the safety story, the "Awaiting QA" label is used
- [x] I have confidence that this PR will not introduce a regression for the reasons below

### Automated test coverage

<!-- Identify the related test coverage and the tests it would catch -->

### QA Plan

<!--
- Describe QA plan that along with automated test coverages proves this PR is regression free
- Link to QA Ticket
-->
No QA
### Safety story
<!--
Describe any other pieces to the safety story including
local testing, why the change is inherently safe, and/or plans to limit the blast radius of a defect.
-->
App manager automated tests.
### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [x] This PR can be reverted after deploy with no further considerations

gherceg committed:
break up import_app to make it more digestible

orangejenny commented on corehq/apps/app_manager/models.py:
```diff
@@ -5707,63 +5707,89 @@ def reapply_overrides(self):
 
 
 def import_app(app_id_or_source, domain, source_properties=None, request=None, check_all_reports=True):
```
Additional possible refactor: it looks like `check_all_reports` is never passed in by any calls to this function.

orangejenny replied:
Looks like the same is true for `source_properties`

gherceg replied:
I think you are right about `check_all_reports` but `source_properties` is passed in here:
https://github.com/dimagi/commcare-hq/blob/9366973aba62ab189848c5e26d30e4d5a0de1af9/corehq/apps/app_manager/views/apps.py#L492
and a bit further (farther?) down in that file. That tricky `...import import_app as import_app_util`

orangejenny replied:
aha, good catch

orangejenny commented on corehq/apps/app_manager/models.py:
```diff
@@ -5707,63 +5707,89 @@ def reapply_overrides(self):
+    if 'build_spec' in source:
+        del source['build_spec']
+
+    app = _copy_app_from_source(domain, source)
```
Realize you're working with existing code that uses the word "source", but I think `doc` is more explicit.

gherceg replied:
I agree that the initially passed in `app_id_or_source` should be renamed to `app_id_or_doc`, but I'm having a hard time understanding what this `source` variable represents in this method because it doesn't seem like it is a doc. It really depends on what this line does `source = source_app.export_json(dump_json=False)` which doesn't give us back a doc right? 

I see `source` useful to describe the source/original/parent doc we are copying from, so it fits when used in `source_app` or `source_domain`. In this case it looks like it would be `source_json` but not sure if that is helpful to future readers.

orangejenny replied:
I'm not sure if we're splitting hairs over what exactly a doc is, but I'd say `export_json` returns a doc (at least when `dump_json` is false, otherwise it returns a string). It calls `to_json` on the app, which is essentially re-generating the doc, and then modifies it slightly.

gherceg replied:
Interesting. So since the `app_id_or_doc` could be either, in the case where that is a doc, are we converting the doc to an `Application` and then back to a doc? Or should I think of the `Application` as a more specific doc? Hopefully that makes sense.

orangejenny replied:
> are we converting the doc to an Application and then back to a doc

Yes. Which sounds weird when said explicitly. The function accepts a source doc for the sake of the import app UI, where the user pastes in a doc. Then that doc is converted to a real app for the sake of being able to call `export_json`, which is needed so that `scrub_source` gets called and unique ids get replaced. Then it gets converted back to a doc because dicts are convenient to work with when you're copying a couch-based class, since you can directly manipulate all the attributes. I don't think it has to be this way - `scrub_source` could be a helper function instead of a method on `Application` - I think it's just how this code evolved.

orangejenny commented on corehq/apps/app_manager/models.py:
```diff
@@ -5707,63 +5707,89 @@ def reapply_overrides(self):
+                if domain_to_add not in media.valid_domains:
+                    media.valid_domains.append(domain_to_add)
                     media.save()
     except ReportConfigurationNotFoundError:
```
I don't see and can't imagine anywhere that a `ReportConfigurationNotFoundError` would be thrown while getting multimedia. I think this try/except can be removed, which would also let you remove the request param from `import_app`.

If you do this, it's worth spending a few minutes blaming this code to try to figure out how this code came to exist.

gherceg replied:
It looks like it is there for good reason based on [this](https://github.com/dimagi/commcare-hq/pull/23991).

orangejenny replied:
Hmm, so the check was deliberately added, but I don't understand what's going on in that PR. Our support for mobile UCR + linked project spaces has changed recently - @mkangia do you have a sense of if this try/catch is still relevant? Even at the time the PR Graham linked was done, I don't see how `get_media_objects` would ever have thrown a `ReportConfigurationNotFoundError`

mkangia replied:
Hey,

The stack trace in this ticket, https://dimagi-dev.atlassian.net/browse/ICDS-240, shares how that exception gets raised when fetching media objects. Looks like its just trying to load modules and hence its configs, ending up complaining about missing reports.

Considering a bit i have known about linked ucrs recently, i imagine, if someone copies an app that has mobile reports or mobile reports with UCRs and has not set linked ucrs, this error message should pop up.
And would be good to know that app would break. Plus, if we don't rescue it, the process if copy would fail.

Does that make sense?

orangejenny replied:
ohhhhh, thank you, I get it now, I hadn't read the stack trace closely enough. Yep, let's leave this try/catch alone.

orangejenny commented:
Looks good to me.

gherceg commented on corehq/apps/app_manager/models.py:
```diff
@@ -5707,63 +5707,89 @@ def reapply_overrides(self):
+    source = source_app.export_json(dump_json=False)
+
+    attachments = _get_attachments(source)
+    source['_attachments'] = {}
```
Any idea why we set `_attachments` to an empty dict after reading from it? This feels like it could use a comment.

orangejenny replied:
I see that have to be moved to the new app, but I don't know why they're cleared from the old app.

gherceg committed:
address jenny's feedback

- remove check_all_reports parameter since no caller uses it
- rename source to doc where it makes sense

gherceg committed:
swap parameters

gherceg commented:
This should not change any existing behavior so I feel good about merging into master without any QA. Just to get another set of eyes on it, @biyeun can you take a look from the perspective of "the code does what it did before". Thanks.

biyeun reviewed (approved):
Looks good. Thanks @gherceg!

gherceg committed:
add doc in names to be explicit
